### DISCUSSION TITLE
Client Libraries

### DISCUSSION THREAD

**Discussion by kiwicopple** [supabase/supabase#5] [Feature Requests]
This issue is for tracking requests/demand for client libraries. Give a comment a thumbs up if you want the connector built.

Please refrain from adding comments unless you are adding a language which isn't already listed.

**Client libraries**

Our client library is modular. Each sub-library is a standalone implementation for a single external system. This is one of the ways we support existing tools.

- **`supabase-{lang}`**: Combines libraries and adds enrichments.
  - `postgrest-{lang}`: Client library to work with [PostgREST](https://github.com/postgrest/postgrest) 
  - `realtime-{lang}`: Client library to work with [Realtime](https://github.com/supabase/realtime)    
  - `gotrue-{lang}`: Client library to work with [GoTrue](https://github.com/netlify/gotrue)

**Comment by kiwicopple** [comment]
Python

Status:

- `postgrest-py`: [https://github.com/supabase-community/postgrest-py](https://github.com/supabase-community/postgrest-py)(community)
- `realtime-py`: https://github.com/supabase-community/realtime-py (community)
- `gotrue-py`: https://github.com/supabase-community/gotrue-py (community)

**Reply by Special-Esmit** [reply]
This is good
im not expert,but it's good

**Comment by kiwicopple** [comment]
Rust

Status:

- `postgrest-rs`: https://github.com/supabase/postgrest-rs (official)
- `realtime-rs`: not started
- `gotrue-rs`: not started

**Reply by bytemunch** [reply]
Very early days, not even close to MVP, but I've started on https://github.com/bytemunch/realtime-rs :)

**Comment by pdavidow** [comment]
Haskell

**Reply by kiwicopple** [reply]
Status:

- `postgrest-hs`: not started
- `realtime-hs`: not started
- `gotrue-hs`: not started

**Comment by fadhlirahim** [comment]
Swift

**Reply by kiwicopple** [reply]
Status:

- `postgrest-swift`: not started
- `realtime-swift`: not started
- `gotrue-swift`: not started

**Reply by satishbabariya** [reply]
@kiwicopple started working on gotrue client for swift here [https://github.com/satishbabariya/gotrue-swift](https://github.com/satishbabariya/gotrue-swift), the client is fully functional, only tests and docs are pending.
will start working on [https://github.com/satishbabariya/supabase-swift](https://github.com/satishbabariya/supabase-swift) and [https://github.com/satishbabariya/storage-swift](https://github.com/satishbabariya/storage-swift) so users at-least can use auth functionalists.

**Comment by fadhlirahim** [comment]
Kotlin

**Reply by kiwicopple** [reply]
Status:

- `postgrest-kotlin`: https://github.com/supabase/postgrest-kt
- `realtime-kotlin`: not started
- `gotrue-kotlin`: https://github.com/supabase/gotrue-kt

**Reply by kevcodez** [reply]
Trying my luck for a gotrue kotlin client here: https://github.com/kevcodez/gotrue-kt - This is a super early proof of concept (just made the initial commit, no tests or anything yet)

**Reply by kevcodez** [reply]
Moved to supabase org https://github.com/supabase/gotrue-kt

**Reply by io-Karim-developer** [reply]
update and add storage api

**Comment by dukex** [comment]
Ruby

**Reply by kiwicopple** [reply]
Status:

- `postgrest-rb`: not started
- `realtime-rb`: not started
- `gotrue-rb`: not started

**Comment by rmznrddn** [comment]
Go

**Reply by kiwicopple** [reply]
Status:

- `postgrest-go`: not started
- `realtime-go`: not started
- `gotrue-go`: not started

**Reply by kevmo314** [reply]
I've started on `postgrest-go` and `realtime-go` here:

https://github.com/doahuddle/postgrest-go
https://github.com/doahuddle/realtime-go

**Reply by kevmo314** [reply]
I'm also guessing that since `gotrue` is written in go, it's a bit nonsensical to write a wrapper library for it :)

**Reply by kevmo314** [reply]
If anyone is interested, I am looking for some help updating the API to match that of postgrest-js and realtime-js. Currently, the API matches the python library, but having the JS API would be better. Additionally, I'd like to create a supabase-go project to wrap the two, but haven't found the time.

**Reply by nedpals** [reply]
@kevmo314 Luckily I made one for my project right now ;) For anyone interested, I made a Supabase client library that currently supports auth and database queries.

Link: https://github.com/nedpals/supabase-go

**Reply by lukecyca** [reply]
I'm looking for a "gotrue-go" library because I want my Go application to communicate with a gotrue server. I will have a look at nedpals/supabase-go.

> it's a bit nonsensical to write a wrapper library for it :)

Is it @kevmo314? Maybe I'm missing something but it seems just as relevant as any of the other ones. How would you recommend that a Go application interact with gotrue?

**Comment by antosubash** [comment]
C#

**Reply by kiwicopple** [reply]
Status:

- `postgrest-csharp`: https://github.com/supabase/postgrest-csharp (community)
- `realtime-csharp`: https://github.com/supabase/realtime-csharp (community)
- `gotrue-csharp`: https://github.com/supabase/gotrue-csharp

**Comment by AlbertoSaezVela** [comment]
Dart

**Reply by kiwicopple** [reply]
Status: Shipped

- `supabase-dart`: - https://github.com/supabase/supabase-dart (official)
  - `postgrest-dart`: https://github.com/supabase/postgrest-dart (official)
  - `realtime-dart`: https://github.com/supabase/realtime-dart (official)
  - `gotrue-dart`: https://github.com/supabase/gotrue-dart (official)

**Reply by amitkhairnar44** [reply]
I'm working on authentication client for Dart. Added email password login along with logout. It's in early stages though.

**Reply by amitkhairnar44** [reply]
Here's the repo : https://github.com/amitkhairnar44/supabase_auth

**Comment by elangora** [comment]
Java

**Reply by kiwicopple** [reply]
Status:

- `postgrest-java`: not started
- `realtime-java`: not started
- `gotrue-java`: not started

**Reply by anshdevops** [reply]
waiting

**Reply by ccleve** [reply]
Java is the first or second most popular programming language in the world. Just sayin'.
https://www.tiobe.com/tiobe-index/

**Reply by maikeffi** [reply]
Am not sure , i may be wrong - But i think it'll take a lot of convincing to make Java world to move away from Spring Boot . But eventually some one will try :)

**Reply by wiverson** [reply]
Spring Boot + supabase would be great. Postgres and Spring Boot already work together very well. The main area of interest would probably be auth - talking to REST services and Postgres is pretty straightforward in Java. If there is interest I could take a look at it.

**Reply by wiverson** [reply]
https://github.com/supabase/gotrue-kt appears to not actually be publishing to Maven Central despite having Mavan coordinates.

**Reply by edgerunner** [reply]
Once upon a time, that was COBOL 😄

**Reply by wiverson** [reply]
Here is my Java + Spring Boot template built on top of Supabase:

https://github.com/ChangeNode/frisson

Currently it uses the Supabase.js official client as the core of the integration. The template shows using Supabase Auth with Spring Security. Covers both social logins as well as flows such as email verification and password reset. Illustrates Thymeleaf user details integration. Doesn't do PostREST as in Java most folks will talk to the DB directly. Stateless - Java sessions are turned off - as it relies on the JWT for user data.

Spring / Spring Boot REST Templates make it really easy to access Supabase APIs. In my experience only the Auth was tricky due to the flows.

**Reply by plexus** [reply]
A Java library would also make it trivial to leverage supbase from other JVM languages (Kotlin, Clojure, Groovy, JRuby, etc.)

**Reply by KuroXI** [reply]
Any progress to supabase-java?

**Reply by MagnusHJensen** [reply]
I should have put a PR in the relevant location, for a storage-java Client library.
https://github.com/supabase-community/.github/pull/25

**Comment by ImreSamu** [comment]
<a name="logo"/>
<div align="center">
<a href="https://julialang.org/" target="_blank">
<img src="https://julialang.org/images/logo_hires.png" alt="Julia Logo" width="210" height="142"></img>
</a>
</div>

**Comment by Immortalin** [comment]
@elangora @fadhlirahim I think Java and Kotlin can be merged under the same umbrella.

**Comment by noahlh** [comment]
[Crystal](https://crystal-lang.org)

**Comment by Immortalin** [comment]
Lua

**Reply by AndreMiras** [reply]
I've started working on a Lua library.
Be gentle it's my first time with Lua :smile: 
It currently has only a very limited support for select queries, but I will be iterating on feature as I see fit/demand.
https://github.com/AndreMiras/postgrest-lua

**Comment by Immortalin** [comment]
F#

**Comment by Immortalin** [comment]
Clojure

**Comment by Immortalin** [comment]
[V](https://github.com/vlang/v)

**Comment by JCKodel** [comment]
Flutter (Dart): I have a 3 million download app that I'm rewriting. Currently, I use a framework for MSSQL I've created (basically, RPC for Stored Procedures and offline-first sync SQLite <-> MSSQL). I will use Supabase, so I can help writting and/or testing a client for Flutter.

**Reply by steinathan** [reply]
Was running around yesterday and found this [https://github.com/supabase/supabase-dart](https://github.com/supabase/supabase-dart)

**Reply by Riz-waan** [reply]
https://github.com/supabase/supabase/discussions/5#discussioncomment-80199

**Comment by mandarvaze** [comment]
Elixir

**Reply by kiwicopple** [reply]
Status:

- `postgrest-ex`: not started
- `realtime-ex`: not started
- `gotrue-ex`: not started

**Reply by J0** [reply]
For my own tracking:
- `postgrest-ex`: https://github.com/J0/postgrestex
- `gotrue-ex`: https://github.com/joshnuss/gotrue-elixir
-  `realtime-ex`: https://github.com/J0/realtime-ex

**Comment by orcchg** [comment]
Android SDK (written in Kotlin in a way compatible with Java)

**Comment by ethanpil** [comment]
PHP

**Comment by kiwicopple** [comment]
Thanks for the tag @mohsenuss91 . Just a reminder:

> Please refrain from adding comments unless you are adding a language which isn't already listed.

This is for good reason :) if you list the same language twice you may end up splitting the +1's.

A small update from the Supabase team - we're excited about the involvement from the community for additional language support. For example, @lqmanh started working on [postgrest-py](https://github.com/supabase/postgrest-py). 

We are working through our alpha feature-set now, and as soon as it's done we will accelerate the community involvement, starting with client libraries. We plan to make community members maintainers of key libraries and then lend support/guidance. If you are considering contributing, we recommend starting with `postgrest-XXX` (of your language choice), as this has the additional benefit of supporting [PostgREST](http://postgrest.org/en/v7.0.0/). 

More updates soon!

**Comment by wingedrhino** [comment]
May I suggest a migration to [gRPC](https://grpc.io/)? It lets you autogenerate clients in multiple languages. By autogenerate, I mean you write service definitions and input/output data structures [as a Protocol Buffers file](https://en.wikipedia.org/wiki/Protocol_Buffers), and then it creates a stub for the service in whatever language you choose to implement the server in. The clients would be 100% autogenerated from this definition file.

There's C, Java, and Go based bindings officially maintained, and community maintained options for [Elixir](https://github.com/elixir-grpc/grpc) and [Rust](https://github.com/stepancheg/grpc-rust). The C binding also powers [C++, Objective-C, Python, Ruby, PHP, Node.js, and other languages that get extended with C](https://grpc.io/docs/languages/), the Java binding also works with Android, plus there's wrappers for other JVM languages like Kotlin. There is also a [Microsoft .NET Core client](https://docs.microsoft.com/en-us/aspnet/core/grpc/), and a [Dart implementation](https://grpc.io/docs/languages/dart/quickstart/) for Flutter apps. Oh and there's [gRPC-Web](https://github.com/grpc/grpc-web) for web clients. There's even a [hipster-friendly GraphQL wrapper now](https://github.com/google/rejoiner).

You can't POSSIBLY hand-write all these for all the services and maintain them yourself!

If the JavaScript client libraries are currently the most mature, it might make sense to write a thin stateless middleware translation layer as a [Node.js gRPC Service](https://grpc.io/docs/languages/node/basics/). You'd then proxy requests to the various backend modules within Supabase. This should make migration easier, because you can eventually switch current services to gRPC and depreciate this proxy, without changing the clients.

**Reply by kiwicopple** [reply]
Hey @wingedrhino - I'm curious if you've used this before? There are equivalent [generators for Open API](https://openapi-generator.tech/), but from my limited experience and research on HN, the generated (client) code is pretty garbage - you get some basic functionality but then you end up modifying it for anything beyond that, at which point you lose the benefit of the generator.

One other important point though - in the short-term it would definitely be easier for Supabase to build from scratch, but one of our core philosophies is to support existing OSS communities, adding our own resources to bring a Firebase-like experience to OSS. Just today I saw [another community](https://pro.tzkt.io/) adopting our libraries to use with PostgREST. This is how open source wins - a rising tide raises all ships. 

> You can't POSSIBLY hand-write all these for all the services and maintain them yourself!

Not ourselves, with the community :) ⚡ I'd still be willing to consider gRPC though, as long as it sticks to the core principal to support other OSS tools

**Reply by wingedrhino** [reply]
> Hey @wingedrhino - I'm curious if you've used this before? There are equivalent [generators for Open API](https://openapi-generator.tech/), but from my limited experience and research on HN, the generated (client) code is pretty garbage - you get some basic functionality but then you end up modifying it for anything beyond that, at which point you lose the benefit of the generator.

I can't say I _used_ it, but I tried it back in 2016 for a startup that tanked before the gRPC code could hit production. So I guess you could say I've _tried_ it?

gRPC itself isn't overly complicated. It generates classes/structs into which you can serialize/deserialize requests, which are sent up and down as protocol buffers. For each method that's been defined, it generates a client function that routes this request through HTTP/2. Even this seems to be written in a pluggable way, so one can switch the transport layer ([QUIC support might land soon!](https://github.com/grpc/grpc/issues/19126)) without modifying much in the server. It even has bi-directional streaming! Perfect for reactive mobile apps (gRPC supports iOS, Android, Ionic, React Native, _and_ Flutter).

Also, using an auto-generated gRPC client library is usually super easy. You initialize it with the address of the server, then make function calls like it were a local method.

You end up having to rewrite things only when you intend to use the classes/structs generated by gRPC for other things and extend them with utility methods. Keep _these_ requirements to a minimum (hopefully zero), and you will not re-write gRPC clients by hand.

Are there any _supabase specific_ helper methods we'd need to implement on all clients? Because a lot of these would be relatively popular functionality (with widely available libraries) like [JWT parsing](https://jwt.io), UUID generation, and JSON parsing.

BTW, a couple of years ago, I tried defining a lisp-like programming language solely as a subset of JSON (I called it LispyJSON), with the idea that it'll be easy to translate that into any other programming language, or write an interpreter for it, and write a visual programming interface, because every language has if/else statements and a JSON parsing library. I think I went as far as writing an interpreter each in Go and JavaScript. Maybe _that_ can be contributed as an extension to gRPC, so that you can define helper methods _on top of_ gRPC service definitions and it'll transpile things for you into client code! :grin: 

> One other important point though - in the short-term it would definitely be easier for Supabase to build from scratch, but one of our core philosophies is to support existing OSS communities, adding our own resources to bring a Firebase-like experience to OSS. Just today I saw [another community](https://pro.tzkt.io/) adopting our libraries to use with PostgREST. This is how open source wins - a rising tide raises all ships.

Yeah. As long as we stick to PostgreSQL and enforce some conventions on top of it, this can be pretty easy to extend. This should be on a different issue, but I'd like us to define a standard-ish way to model data on top of PostgreSQL. Because if PostgreSQL is an integration point, you can have more than one application that points to it directly, each serving a different purpose.

Say you're happy with Postgrest for a RESTful API. But you might want to include [Hasura](https://hasura.io/) for its GraphQL server. And perhaps you'd like to use [Metabase](metabase.com/) for an internal analytics dashboard (it auto-generates some stats and charts by examining your schema. Try it if you haven't already!). Or maybe there's an opensource AirTable/Access style no-code programming interface that comes up in the future, and it too points to PostgreSQL.

It'll be fab if _everything can play well with the same PostgreSQL instance_, since expecting everyone to go through ONE middleware is near impossible.

PostgreSQL is a great investment. At some point in the future, there'll be enough feature parity between [CitusData](citusdata.com/), [YugabyteDB](https://www.yugabyte.com/), and [CockroachDB](https://www.cockroachlabs.com/), making Supabase scale _very_ nicely.

> > You can't POSSIBLY hand-write all these for all the services and maintain them yourself!
> 
> Not ourselves, with the community :) zap I'd still be willing to consider gRPC though, as long as it sticks to the core principal to support other OSS tools

I used _you_ in plural form lol. Even as a community, maintaining hand-written drivers for multiple languages is hard. Many niche non-relational databases struggle keeping feature parity across all the language clients.

gRPC being gRPC, `popular language X` from the future will have a gRPC implementation for it. This is great for improving adoption! Plus, gRPC keeps getting features added to it (it too is an opensource project) so we'll definitely see gains from it in the future.

Finally, I have one last pitch for auto-generated API clients. Every minute spent by a human doing work that could've been done by a computer, there's a puppy somewhere that doesn't have anyone to play with. :yum:

**Reply by kiwicopple** [reply]
Thanks for the in-depth reply @wingedrhino - perhaps you can shift this your top level comment to a new Discussion, then we can shift the convo over? I don't want to sidetrack this discussion and I'm aware there are a lot of subscribers who might just be interested in getting notified of new releases

**Comment by smorimoto** [comment]
OCaml

**Comment by carlosstenzel** [comment]
JavaScript

**Reply by soedirgo** [reply]
Hey @carlosstenzel! We already have a JS client here: https://github.com/supabase/supabase-js

**Reply by carlosstenzel** [reply]
I already seen, I commented so that it remains the official :D

**Comment by klutchdev** [comment]
So many possibilities!!! 

I'd love to see a client library for Flutter/Dart, which not only compiles to JavaScript &&  native code for iOS/Android - but is also very similar to Javascript/React regarding the implementation of a back end and functional components. 

With Flutter/Dart and JavaScript having official support, Supabase would be able to provide its glorious magic to just about every device and operating system on the planet.

My runner up choice is any sort of potential support for C/CPP, which can be paired up with React to build awesome Web Assembly stuff!

**Reply by Riz-waan** [reply]
https://github.com/supabase/supabase/discussions/5#discussioncomment-80199

**Comment by jdlynn** [comment]
# Python

**Reply by ryanhaticus** [reply]
https://github.com/supabase/supabase/discussions/5#discussioncomment-80191

**Comment by nailuj29** [comment]
Elm

**Comment by gigabytejayesh** [comment]
**Arduino library**

_Board support: ESP's, UNO board, etc_

**Reply by Iccr** [reply]
are there any library for arduino currently?

**Reply by wvffle** [reply]
I want to ask the same question a year later, are there any libraries now?

**Comment by abnemo** [comment]
Deno

Just in case. I am not sure, maybe the JS client is already on the way to get deno compatibility. =)

**Reply by beeequeue** [reply]
Since the libraries are intended to be possible to use in browsers it shouldn't be dependent on any node code, if so it _should_ 🤞 work with deno afaik

**Comment by gideongrinberg** [comment]
Go(lang)

**Reply by nedpals** [reply]
Check this [one](https://github.com/supabase/supabase/discussions/5#discussioncomment-80197). :)

**Reply by gideongrinberg** [reply]
My bad, thanks for pointing it out :)

**Comment by arnu515** [comment]
Deno!

It should be pretty easy, since `denoify` exists

**Reply by Riz-waan** [reply]
https://github.com/supabase/supabase/discussions/5#discussioncomment-639968

**Comment by zorreth** [comment]
Java

**Comment by maxbaluev** [comment]
Julia

**Comment by grdsdev** [comment]
How we make sure the client libs shares the same api between clients?
Why don't write a client lib generator? 
Something like [swagger-codegen](https://github.com/swagger-api/swagger-codegen)

**Reply by soedirgo** [reply]
This is a big limit on flexibility. All languages have idiosyncrasies - I think we should allow client libraries to deviate a little to match the idiomatic style for the respective languages, as long as it covers the core functionality.

**Comment by roker15** [comment]
TypeScript -

**Reply by soedirgo** [reply]
https://github.com/supabase/supabase-js already supports TypeScript @roker15

**Reply by roker15** [reply]
I tried to do following in my nextjs - typescript  project but it did not work
`import { createClient } from '@supabase/supabase-js`
I had to find workaround by adding `declare module "@supabase/supabase-js` in ` .d.ts ` file . So it's not working out of the box.  could You add typescript version of of demo example in your documentation for next js

**Reply by soedirgo** [reply]
Here's a TS Next.js example: https://github.com/supabase/supabase/tree/master/examples/nextjs-ts-user-management

**Comment by lunandd** [comment]
Clojure

**Reply by plexus** [reply]
Note that a Java client would be good enough to use it in Clojure projects.

**Comment by yob-yob** [comment]
php

**Reply by Nilanth** [reply]
@kiwicopple FYA:

**Comment by PeggyWang0** [comment]
C++

**Comment by lessp** [comment]
OCaml 🐫

**Comment by tagorenathv** [comment]
@kiwicopple Is there a Java library to use with Spring Boot?

**Reply by wiverson** [reply]
FWIW there are libs posted at https://github.com/supabase-community for GoTrue and Storage.

In practice you can just connect your Spring Boot app via JDBC and go for most server side stuff. On the client you'll likely want to just use the JS client and then send the auth JWT back to the server. I did some work on the C# client for GoTrue to add in a bunch of fancier stuff eg offline, network, auth sessions etc but if you are just using web clients IMHO just use the Supabase JS client for auth etc.

TBH using Spring Boot for backend REST services and SvelteKit in SPA mode for the client has been excellent, including deploying to Vercel and also to mobile via Capacitor...
